store/store_delete.go

```diff
@@ -348,3 +230,186 @@ func (s *Store[H]) deleteParallel(ctx context.Context, from, to uint64) (uint64,
 	)
 	return highest, missing, nil
 }
+
+// DeleteRange deletes headers in the range [from:to) from the store.
+// It intelligently updates head and/or tail pointers based on what range is being deleted.
+func (s *Store[H]) DeleteRange(ctx context.Context, from, to uint64) error {
+	// ensure all the pending headers are synchronized
+	err := s.Sync(ctx)
+	if err != nil {
+		return err
+	}
+
+	head, err := s.Head(ctx)
+	if err != nil {
+		return fmt.Errorf("header/store: reading head: %w", err)
+	}
+
+	tail, err := s.Tail(ctx)
+	if err != nil {
+		return fmt.Errorf("header/store: reading tail: %w", err)
+	}
+
+	// validate range parameters
+	if from >= to {
+		return fmt.Errorf(
+			"header/store: invalid range [%d:%d) - from must be less than to",
+			from,
+			to,
+		)
+	}
+
+	if from < tail.Height() {
+		return fmt.Errorf(
+			"header/store: delete range from %d below current tail(%d)",
+			from,
+			tail.Height(),
+		)
+	}
+
+	// Note: Allow deletion beyond head to match original DeleteTo behavior
+	// Missing headers in the range will be handled gracefully by the deletion logic
+
+	// if range is empty within the current store bounds, it's a no-op
+	if from > head.Height() || to <= tail.Height() {
+		return nil
+	}
+
+	// Validate that deletion won't create gaps in the store
+	// Only allow deletions that:
+	// 1. Start from tail (advancing tail forward)
+	// 2. End at head+1 (moving head backward)
+	// 3. Delete the entire store
+	if from > tail.Height() && to <= head.Height() {
+		return fmt.Errorf(
+			"header/store: deletion range [%d:%d) would create gaps in the store. "+
+				"Only deletion from tail (%d) or to head+1 (%d) is supported",
+			from, to, tail.Height(), head.Height()+1,
+		)
+	}
+
+	// Check if we're deleting all existing headers (making store empty)
+	// Only wipe if 'to' is exactly at head+1 (normal case) to avoid accidental wipes
+	if from <= tail.Height() && to == head.Height()+1 {
```

Comment:
do we allow a case where caller specifies a `from` that is < `tail.Height()` ? I don't see why we shouldn't but just want to make sure this isn't some weirdness we should silently overlook.

Reply:
L262 we were already verifying that. We can indeed just do an equality check for from.